src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.cs

```diff
@@ -111,6 +112,16 @@ private bool WaitOneNoCheck(int millisecondsTimeout)
             SafeWaitHandle? waitHandle = _waitHandle;
             ObjectDisposedException.ThrowIf(waitHandle is null, this);
 
+            bool isWaitHandleKeywordEnabled = NativeRuntimeEventSource.Log.IsEnabled(
+                EventLevel.Informational,
+                NativeRuntimeEventSource.Keywords.WaitHandleKeyword);
+            if (isWaitHandleKeywordEnabled)
+            {
+                NativeRuntimeEventSource.Log.WaitHandleWaitStart(
```

Comment:
I think there are cases where duplicate events would be raised for the same wait. For instance in NativeAOT, Condition would raise an event, and another event would be raised here. In CoreCLR, this event would be raised and another in Thread::DoAppropriateWait. There's probably some tweaking to do. For NativeAOT/Mono, I think raising the event from the Wait*Core methods would work, along with passing the info down through internal methods. For CoreCLR, it just let the native side raise the event.

Reply:
Hm it's very messy in my head, I'll try drawing the different paths and get back with a proposal.

Reply:
![image](https://github.com/dotnet/runtime/assets/9092290/c526b2d1-5a52-41b0-897a-a7377a6809b9)

If that's accurate is it needed to emit an event in managed code? 🤔

Reply:
Thinking about it more, it seems Monitor.Wait is kind of a special case where it's implemented in native code in CoreCLR and Mono, and almost always would be a blocking wait. For all other WaitHandle.Wait* cases, it would probably make sense to raise the event from the managed implementations. Maybe it would make sense to special-case the raising of the event for Monitor.Wait (though in Mono it looks like it could also be raised in managed code before calling out to the native implementation). I'll try to draft up a change of what I'm thinking.

Reply:
May need to also check in CoreCLR other wait cases from the native side that lead to waiting on wait handles, to see if it would be useful to cover.

Reply:
Yea probably makes sense to raise the event in Thread::DoAppropriateWait or below in CoreCLR. There are a couple other tweaks I was thinking of, will try to get a draft out sometime soon.

Reply:
I've removed all managed code from the PR

Reply:
It looks like with the latest changes it would only work in CoreCLR.

I made some changes in this commit: https://github.com/dotnet/runtime/commit/35c13cfe05578ad2fabfe6711dcdd2ae1be0dcec

That brings back the minimal code to raise the event from the managed side for NativeAOT/Mono, along with a few other changes. The commit description has additional details. I've tested it and the events appear to work in CoreCLR, NativeAOT, and Mono. Please take a look, if that looks ok to you, maybe you can merge it into your branch?

Reply:
Doesn't NativeAOT use the native code where the event emitted from? The test is passing without disabling NativeAOT.

I'm a bit confused about the mono part though, where is the event emitted?

Reply:
For WaitHandle, NativeAOT and Mono on Windows [p/invoke into Windows APIs](https://github.com/dotnet/runtime/blob/1aae18a21ebb5f08a2a734cfe31ba4bd00b2ad7b/src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs#L52-L61), and on Unixes use the [managed WaitSubsystem implementation](https://github.com/dotnet/runtime/blob/1aae18a21ebb5f08a2a734cfe31ba4bd00b2ad7b/src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.Unix.cs#L11).

Monitor.Wait on Mono has its implementation in native code, the event is emitted here in my commit: https://github.com/dotnet/runtime/commit/35c13cfe05578ad2fabfe6711dcdd2ae1be0dcec#diff-847bb36cd91f623c5b9f55a1980656052b342e6d93da972a41ded1210b57d719R130

Reply:
> The test is passing without disabling NativeAOT.

The test currently doesn't run with NativeAOT because the RemoteExecutor is currently not supported for it.